File: dmn-test-cases/standard/tck/1.1/cl3/0031-static-user-defined-functions/translator/expected/java/dmn/FnLibrary.java

import java.util.*;
import java.util.stream.Collectors;

@javax.annotation.Generated(value = {"decision.ftl", "fn library"})
@com.gs.dmn.runtime.annotation.DRGElement(
    namespace = "",
    name = "fn library",
    label = "",
    elementKind = com.gs.dmn.runtime.annotation.DRGElementKind.DECISION,
    expressionKind = com.gs.dmn.runtime.annotation.ExpressionKind.CONTEXT,
    hitPolicy = com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
    rulesCount = -1
)
public class FnLibrary extends com.gs.dmn.runtime.JavaTimeDMNBaseDecision<type.TFnLibrary> {
    public static final com.gs.dmn.runtime.listener.DRGElement DRG_ELEMENT_METADATA = new com.gs.dmn.runtime.listener.DRGElement(
        "",
        "fn library",
        "",
        com.gs.dmn.runtime.annotation.DRGElementKind.DECISION,
        com.gs.dmn.runtime.annotation.ExpressionKind.CONTEXT,
        com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
        -1
    );

    public FnLibrary() {
    }

    @java.lang.Override()
    public type.TFnLibrary applyMap(java.util.Map<String, String> input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return apply(context_);
        } catch (Exception e) {
            logError("Cannot apply decision 'FnLibrary'", e);
            return null;
        }
    }

    @java.lang.Override()
    public type.TFnLibrary applyPojo(com.gs.dmn.runtime.ExecutableDRGElementInput input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return apply(context_);
        } catch (Exception e) {
            logError("Cannot apply element 'FnLibrary'", e);
            return null;
        }
    }

    @java.lang.Override()
    public type.TFnLibrary applyContext(com.gs.dmn.runtime.Context input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return applyPojo(new FnLibraryInput_(input_), context_);
        } catch (Exception e) {
            logError("Cannot apply element 'FnLibrary'", e);
            return null;
        }
    }

    public type.TFnLibrary apply(com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            // Start decision 'fn library'
            com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
            com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
            com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
            com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;
            long fnLibraryStartTime_ = System.currentTimeMillis();
            com.gs.dmn.runtime.listener.Arguments fnLibraryArguments_ = new com.gs.dmn.runtime.listener.Arguments();
            eventListener_.startDRGElement(DRG_ELEMENT_METADATA, fnLibraryArguments_);

            // Evaluate decision 'fn library'
            type.TFnLibrary output_ = lambda.apply(context_);

            // End decision 'fn library'
            eventListener_.endDRGElement(DRG_ELEMENT_METADATA, fnLibraryArguments_, output_, (System.currentTimeMillis() - fnLibraryStartTime_));

            return output_;
        } catch (Exception e) {
            logError("Exception caught in 'fn library' evaluation", e);
            return null;
        }
    }

    public com.gs.dmn.runtime.LambdaExpression<type.TFnLibrary> lambda =
        new com.gs.dmn.runtime.LambdaExpression<type.TFnLibrary>() {
            public type.TFnLibrary apply(Object... args_) {
                com.gs.dmn.runtime.ExecutionContext context_ = 0 < args_.length ? (com.gs.dmn.runtime.ExecutionContext) args_[0] : null;
                com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
                com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
                com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
                com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;

                com.gs.dmn.runtime.LambdaExpression<java.lang.Number> sumFn = new com.gs.dmn.runtime.LambdaExpression<java.lang.Number>() {public java.lang.Number apply(Object... args_) {java.lang.Number a = (java.lang.Number)args_[0]; java.lang.Number b = (java.lang.Number)args_[1];return numericAdd(a, b);}};
                com.gs.dmn.runtime.LambdaExpression<java.lang.Number> subFn = new com.gs.dmn.runtime.LambdaExpression<java.lang.Number>() {public java.lang.Number apply(Object... args_) {java.lang.Number a = (java.lang.Number)args_[0]; java.lang.Number b = (java.lang.Number)args_[1];return numericSubtract(a, b);}};
                com.gs.dmn.runtime.LambdaExpression<java.lang.Number> multiplyFn = new com.gs.dmn.runtime.LambdaExpression<java.lang.Number>() {public java.lang.Number apply(Object... args_) {java.lang.Number a = (java.lang.Number)args_[0]; java.lang.Number b = (java.lang.Number)args_[1];return numericMultiply(a, b);}};
                com.gs.dmn.runtime.LambdaExpression<java.lang.Number> divideFn = new com.gs.dmn.runtime.LambdaExpression<java.lang.Number>() {public java.lang.Number apply(Object... args_) {java.lang.Number a = (java.lang.Number)args_[0]; java.lang.Number b = (java.lang.Number)args_[1];return (booleanEqual(numericEqual(b, number("0")), Boolean.TRUE)) ? null : numericDivide(a, b);}};
                type.TFnLibraryImpl fnLibrary = new type.TFnLibraryImpl();
                fnLibrary.setSumFn(sumFn);
                fnLibrary.setSubFn(subFn);
                fnLibrary.setMultiplyFn(multiplyFn);
                fnLibrary.setDivideFn(divideFn);
                return fnLibrary;
            }
        };
}

File: dmn-test-cases/standard/tck/1.1/cl3/0031-static-user-defined-functions/translator/expected/kotlin/dmn/FnLibrary.kt

import java.util.*
import java.util.stream.Collectors

@javax.annotation.Generated(value = ["decision.ftl", "fn library"])
@com.gs.dmn.runtime.annotation.DRGElement(
    namespace = "http://www.actico.com/spec/DMN/0.1.0/0031-user-defined-functions",
    name = "fn library",
    label = "",
    elementKind = com.gs.dmn.runtime.annotation.DRGElementKind.DECISION,
    expressionKind = com.gs.dmn.runtime.annotation.ExpressionKind.CONTEXT,
    hitPolicy = com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
    rulesCount = -1
)
class FnLibrary() : com.gs.dmn.runtime.JavaTimeDMNBaseDecision<type.TFnLibrary?>() {
    override fun applyMap(input_: MutableMap<String, String>, context_: com.gs.dmn.runtime.ExecutionContext): type.TFnLibrary? {
        try {
            return apply(context_)
        } catch (e: Exception) {
            logError("Cannot apply decision 'FnLibrary'", e)
            return null
        }
    }

    fun apply(context_: com.gs.dmn.runtime.ExecutionContext): type.TFnLibrary? {
        try {
            // Start decision 'fn library'
            var annotationSet_: com.gs.dmn.runtime.annotation.AnnotationSet = context_.getAnnotations()
            var eventListener_: com.gs.dmn.runtime.listener.EventListener = context_.getEventListener()
            var externalExecutor_: com.gs.dmn.runtime.external.ExternalFunctionExecutor = context_.getExternalFunctionExecutor()
            var cache_: com.gs.dmn.runtime.cache.Cache = context_.getCache()
            val fnLibraryStartTime_ = System.currentTimeMillis()
            val fnLibraryArguments_ = com.gs.dmn.runtime.listener.Arguments()
            eventListener_.startDRGElement(DRG_ELEMENT_METADATA, fnLibraryArguments_)

            // Evaluate decision 'fn library'
            val output_: type.TFnLibrary? = evaluate(context_)

            // End decision 'fn library'
            eventListener_.endDRGElement(DRG_ELEMENT_METADATA, fnLibraryArguments_, output_, (System.currentTimeMillis() - fnLibraryStartTime_))

            return output_
        } catch (e: Exception) {
            logError("Exception caught in 'fn library' evaluation", e)
            return null
        }
    }

    private inline fun evaluate(context_: com.gs.dmn.runtime.ExecutionContext): type.TFnLibrary? {
        var annotationSet_: com.gs.dmn.runtime.annotation.AnnotationSet = context_.getAnnotations()
        var eventListener_: com.gs.dmn.runtime.listener.EventListener = context_.getEventListener()
        var externalExecutor_: com.gs.dmn.runtime.external.ExternalFunctionExecutor = context_.getExternalFunctionExecutor()
        var cache_: com.gs.dmn.runtime.cache.Cache = context_.getCache()
        val sumFn: com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>? = com.gs.dmn.runtime.LambdaExpression<kotlin.Number> { args_ -> val a: kotlin.Number? = args_[0] as kotlin.Number?; val b: kotlin.Number? = args_[1] as kotlin.Number?; numericAdd(a, b) } as com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>?
        val subFn: com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>? = com.gs.dmn.runtime.LambdaExpression<kotlin.Number> { args_ -> val a: kotlin.Number? = args_[0] as kotlin.Number?; val b: kotlin.Number? = args_[1] as kotlin.Number?; numericSubtract(a, b) } as com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>?
        val multiplyFn: com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>? = com.gs.dmn.runtime.LambdaExpression<kotlin.Number> { args_ -> val a: kotlin.Number? = args_[0] as kotlin.Number?; val b: kotlin.Number? = args_[1] as kotlin.Number?; numericMultiply(a, b) } as com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>?
        val divideFn: com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>? = com.gs.dmn.runtime.LambdaExpression<kotlin.Number> { args_ -> val a: kotlin.Number? = args_[0] as kotlin.Number?; val b: kotlin.Number? = args_[1] as kotlin.Number?; (if (booleanEqual(numericEqual(b, number("0")), true)) null else numericDivide(a, b)) } as com.gs.dmn.runtime.LambdaExpression<kotlin.Number?>?
        val fnLibrary: type.TFnLibraryImpl? = type.TFnLibraryImpl() as type.TFnLibraryImpl?
        fnLibrary?.sumFn = sumFn
        fnLibrary?.subFn = subFn
        fnLibrary?.multiplyFn = multiplyFn
        fnLibrary?.divideFn = divideFn
        return fnLibrary
    }

    companion object {
        val DRG_ELEMENT_METADATA : com.gs.dmn.runtime.listener.DRGElement = com.gs.dmn.runtime.listener.DRGElement(
            "",
            "fn library",
            "",
            com.gs.dmn.runtime.annotation.DRGElementKind.DECISION,
            com.gs.dmn.runtime.annotation.ExpressionKind.CONTEXT,
            com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
            -1
        )
    }
}
